WindowsAzureMessaging/WindowsAzureMessaging/SBNotificationHub.m

```diff
@@ -378,19 +378,65 @@ - (void)deleteRegistrationWithName:(NSString *)templateName completion:(void (^)
 - (void)unregisterAllWithDeviceToken:(NSData *)deviceTokenData completion:(void (^)(NSError *))completion {
     if (deviceTokenData == nil) {
         if (completion) {
-            completion([SBNotificationHubHelper errorForNullDeviceToken]);
+            dispatch_async(dispatch_get_main_queue(), ^{
+                completion([SBNotificationHubHelper errorForNullDeviceToken]);
+            });
         }
 
         return;
     }
 
     NSString *deviceToken = [self convertDeviceToken:deviceTokenData];
+    
```

Comment:
Let me make sure I understand the change. We are collapsing the logic from `deleteAllRegistrationThread:` into this method?

It looks like this method performs all operations on the calling thread now, but always ships completions back to the main thread. Is that correct?

Reply:
Correct.  Previously, it was going to the background thread because all of the calls it was making for `uregisterAll:` were blocking calls.  This PR eliminates the blocking calls and replaces them with callbacks.  In order to do the for loop with callbacks is with using a `dispatch_group_t`.  To be consistent then for all callbacks and the join for the dispatch group, I needed to pick a thread, so the main thread was the ideal spot.

Reply:
Are we sure each "async" call in this case is non-blocking? I'm hesitant to assume the presence of a completion callback means it's 100% async.

Reply:
Yes, each call is non-blocking as it goes all the way down to the `SBURLConnection` and the `sendRequest` method, it is completely non-blocking versus the other calls that have an error out value call the `sendSynchronousRequest` method.

Reply:
Okay, great. Thanks for confirming!

What about always calling the completion callback on the main thread? Are we assuming that calls always originate on the main thread?